rust/trace-dump/src/chrome_trace.rs

```diff
@@ -29,58 +26,175 @@ pub enum OutputFormat {
 #[derive(Debug)]
 pub enum Error {
     Io(IOError),
-    Json(serde_json::Error)
+    Json(serde_json::Error),
+    DecodingFormat(String)
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct ChromeTraceArrayEntryArgs {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    xi_sample_id: Option<usize>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    payload: Option<TracePayloadT>,
+}
+
+/// This is the struct that represents a single entry within an array of
+/// samples.  Vec<ChromeTraceArrayEntry> would represent a single complete
+/// trace.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct ChromeTraceArrayEntry {
+    name: StrCow,
+    cat: StrCow,
+    ph: StrCow,
+    #[serde(rename = "ts")]
+    ts_us: u64,
+    pid: u64,
+    tid: u64,
+    args: Option<ChromeTraceArrayEntryArgs>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(untagged)]
+enum ChromeTraceArrayEntries {
+    Array(Vec<ChromeTraceArrayEntry>)
 }
 
 fn to_us(ns: u64) -> u64 {
     ns / 1000
 }
 
+fn to_ns(us: u64) -> u64 {
+    us * 1000
+}
+
 fn event_type(sample: &Sample, begin: bool) -> &'static str {
     match sample.sample_type {
         SampleType::Instant => "i",
         SampleType::Duration => if begin { "B" } else { "E" },
     }
 }
 
-fn sample_to_json(sample: &Sample, begin: bool) -> serde_json::Value {
-    json!({
-        "cat": sample.categories.join(","),
-        "name": sample.name, 
-        "ph": event_type(sample, begin),
-        "ts": to_us(if begin { sample.start_ns } else { sample.get_end_ns() }),
-        "pid": sample.pid,
-        "tid": sample.tid,
-        "args": {
-            "payload": sample.payload
+fn sample_to_entry(sample: &Sample, begin: bool) -> ChromeTraceArrayEntry {
+    ChromeTraceArrayEntry {
+        name: sample.name.clone(),
+        cat: StrCow::from(sample.categories.join(",")),
```

Comment:
Thinking about this being a `Cow`... I guess the borrow case is if there's one category and it's a static &'str, we can skip the allocation?

I'm actually confused about `join` being available here, in the docs I have it seems to be nightly only?